internal/daemon/api_files.go

```diff
@@ -0,0 +1,627 @@
+	fileTypeSocket    fileType = "socket"
+	fileTypeNamedPipe fileType = "named-pipe"
+	fileTypeDevice    fileType = "device"
+	fileTypeUnknown   fileType = "unknown"
+)
+
+func fileModeToType(mode os.FileMode) fileType {
+	switch {
+	case mode&os.ModeType == 0:
+		return fileTypeFile
+	case mode&os.ModeDir != 0:
+		return fileTypeDirectory
+	case mode&os.ModeSymlink != 0:
+		return fileTypeSymlink
+	case mode&os.ModeSocket != 0:
+		return fileTypeSocket
+	case mode&os.ModeNamedPipe != 0:
+		return fileTypeNamedPipe
+	case mode&os.ModeDevice != 0:
+		return fileTypeDevice
+	default:
+		return fileTypeUnknown
+	}
+}
+
+func fileInfoToResult(fullPath string, info os.FileInfo) fileInfoResult {
+	mode := info.Mode()
+	var psize *int64
+	if mode.IsRegular() {
+		size := info.Size()
+		psize = &size
+	}
+	result := fileInfoResult{
+		Path:         fullPath,
+		Name:         path.Base(fullPath),
+		Type:         fileModeToType(mode),
+		Size:         psize,
+		Permissions:  fmt.Sprintf("%03o", mode.Perm()),
+		LastModified: info.ModTime().Format(time.RFC3339),
+	}
+	return result
+}
+
+func listFiles(pattern string, directoryItself bool) Response {
+	if !path.IsAbs(pattern) {
+		return statusBadRequest("pattern must be absolute (%q is not)", pattern)
+	}
+	result, err := listFilesErr(pattern, directoryItself)
+	if err != nil {
+		return listErrorResponse(err)
+	}
+	return SyncResponse(result)
+}
+
+func listFilesErr(pattern string, directoryItself bool) ([]fileInfoResult, error) {
+	info, err := os.Stat(pattern)
+	if errors.Is(err, os.ErrNotExist) {
+		dir, base := filepath.Split(pattern)
+		result, err := readDirFiltered(dir, base)
+		if err != nil {
+			return nil, err
+		}
+		if len(result) == 0 {
+			// They specified a file or pattern and it doesn't exist.
+			return nil, os.ErrNotExist
+		}
+		return result, nil
+	}
+	if err != nil {
+		return nil, err
+	}
+	if !info.IsDir() || directoryItself {
+		// Info about a single file (or directory entry itself).
+		result := []fileInfoResult{fileInfoToResult(pattern, info)}
+		return result, nil
+	}
+	return readDirFiltered(pattern, "")
+}
+
+func readDirFiltered(dir, pattern string) ([]fileInfoResult, error) {
+	infos, err := ioutil.ReadDir(dir)
+	if err != nil {
+		return nil, err
+	}
+	var result []fileInfoResult
+	for _, info := range infos {
+		name := info.Name()
+		matched := true
+		if pattern != "" {
+			matched, err = filepath.Match(pattern, name)
+		}
+		if matched {
+			p := filepath.Join(dir, name)
+			result = append(result, fileInfoToResult(p, info))
+		}
+	}
+	return result, nil
+}
+
+func v1PostFiles(_ *Command, req *http.Request, _ *userState) Response {
+	contentType := req.Header.Get("Content-Type")
+	mediaType, params, err := mime.ParseMediaType(contentType)
+	if err != nil {
+		return statusBadRequest("invalid Content-Type %q", contentType)
+	}
+
+	switch mediaType {
+	case "multipart/form-data":
+		boundary := params["boundary"]
+		if boundary == "" {
```

Comment:
We should probably do a basic check here as well to make sure this is not too short, otherwise a cooked result might end up creating a disproportional amount of work on the server end.

Reply:
Done. I've added `const minBoundaryLength 8`.